python-squid/src/main/java/org/sonar/python/metrics/CognitiveComplexityVisitor.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.metrics;
+
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.AstNodeType;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.sonar.python.PythonVisitor;
+import org.sonar.python.api.PythonGrammar;
+import org.sonar.python.api.PythonKeyword;
+import org.sonar.python.api.PythonPunctuator;
+
+public class CognitiveComplexityVisitor extends PythonVisitor {
+
+  private AstNode currentFunction = null;
+  private int complexity = 0;
+  private int nestingLevel = 0;
+  @Nullable
+  private final SecondaryLocationConsumer secondaryLocationConsumer;
+
+  public interface SecondaryLocationConsumer {
+    void consume(AstNode node, String message);
+  }
+
+  CognitiveComplexityVisitor(@Nullable SecondaryLocationConsumer secondaryLocationConsumer) {
+    this.secondaryLocationConsumer = secondaryLocationConsumer;
+  }
+
+  public static int complexity(AstNode node, @Nullable SecondaryLocationConsumer secondaryLocationConsumer) {
+    CognitiveComplexityVisitor visitor = new CognitiveComplexityVisitor(secondaryLocationConsumer);
+    visitor.scanNode(node);
+    return visitor.complexity;
+  }
+
+  public int getComplexity() {
+    return complexity;
+  }
+
+  @Override
+  public Set<AstNodeType> subscribedKinds() {
+    return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+      PythonGrammar.IF_STMT,
+      PythonKeyword.ELIF,
+      PythonKeyword.ELSE,
+
+      PythonGrammar.WHILE_STMT,
+      PythonGrammar.FOR_STMT,
+      PythonGrammar.EXCEPT_CLAUSE,
+
+      PythonGrammar.AND_TEST,
+      PythonGrammar.OR_TEST,
+
+      PythonGrammar.TEST,
+
+      PythonGrammar.FUNCDEF,
+      PythonGrammar.SUITE)));
+  }
+
+  @Override
+  public void visitNode(AstNode astNode) {
+    if (astNode.is(PythonGrammar.FUNCDEF) && currentFunction == null) {
+      currentFunction = astNode;
+      nestingLevel = 0;
```

Comment:
Here we will lose the previous nesting level when exiting the function, which might be result in wrong computation in case of multiple nested functions.
Ex.
```
def a():
  def b():
    def c():
      if 1 > 0:
        2
    return c() if 1 else "g"
  return b() if 1 > 0 else "f"
```

Reply:
It's not wrong because we only compute complexity of the first level of functions